File: fifo.rb
#!/usr/bin/ruby1.8
# avr-vs1002
# Streaming from serial port to VS1002.
#
# (c) Karl-Martin Skontorp <kms@skontorp.net> ~ http://22pf.org/
# Licensed under the GNU GPL 2.0 or later.

b = Array.new
bs = 1000

f = File.open(ARGV[0])

t1 = Time.now
t2 = Time.now

while true
    begin
        r = f.read(bs)
        if (r != nil) 
            b.push(r)
        end
        if (STDIN.read_nonblock(1) == '.')
            d = t1 - t2

            STDERR.write("%.2fkb/s [%dB]\n" % [((bs * 8) / d / 1000), b.length * bs])

            print b.shift
            STDOUT.flush

            t2 = t1
            t1 = Time.now
        end
    rescue Errno::EAGAIN => e
    end
end

File: fifo.c
/* avr-vs1002
 * Streaming from serial port to VS1002.
 *
 * (c) Karl-Martin Skontorp <kms@skontorp.net> ~ http://picofarad.org/
 * Licensed under the GNU GPL 2.0 or later.
 */

#include <stdlib.h>
#include "fifo.h"

uint8_t isFifoEmpty(const fifo *c) {
    return c->unconsumed == 0;
}

uint8_t fifoSize(const fifo *c) {
    return c->unconsumed;
}

uint8_t fifoGet(fifo *c) {
    c->read = c->read & c->capacity;
    c->unconsumed--;
    return c->buffer[c->read++];
}       

void fifoPut(fifo *c, const uint8_t p) {
    if (c->unconsumed > c->capacity) {
        return;
    }
    c->write = c->write & c->capacity;
    c->buffer[c->write++] = p;
    c->unconsumed++;
}

fifo* fifoCreate(const uint8_t capacity) {
    fifo *f = malloc(sizeof(fifo) + (capacity * sizeof(uint8_t)));

    f->read = 0;
    f->write = 0;
    f->unconsumed = 0;
    f->capacity = capacity - 1;

    return f;
}   
